README.md
> A lightweight, GPU accelerated, SQL engine built on the [RAPIDS.ai](https://rapids.ai) ecosystem.

<a href='https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/welcome.ipynb'>Get Started on app.blazingsql.com</a>

[Getting Started](#getting-started) | [Documentation](https://docs.blazingdb.com) | [Examples](#examples) | [Contributing](#contributing) | [License](LICENSE) | [Blog](https://blog.blazingdb.com) | [Try Now](https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/welcome.ipynb)

BlazingSQL is a GPU accelerated SQL engine built on top of the RAPIDS ecosystem. RAPIDS is based on the [Apache Arrow](http://arrow.apache.org) columnar memory format, and [cuDF](https://github.com/rapidsai/cudf) is a GPU DataFrame library for loading, joining, aggregating, filtering, and otherwise manipulating data.

BlazingSQL is a SQL interface for cuDF, with various features to support large scale data science workflows and enterprise datasets.
* **Query Data Stored Externally** - a single line of code can register remote storage solutions, such as Amazon S3.
* **Simple SQL** - incredibly easy to use, run a SQL query and the results are GPU DataFrames (GDFs).
* **Interoperable** - GDFs are immediately accessible to any [RAPIDS](htts://github.com/rapidsai) library for data science workloads.

Try our 5-min [Welcome Notebook](https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/welcome.ipynb) to start using BlazingSQL and RAPIDS AI.

# Getting Started

Here's two copy + paste reproducable BlazingSQL snippets, keep scrolling to find [example Notebooks](#examples) below.

Create and query a table from a `cudf.DataFrame` with progress bar:

```python
import cudf

df = cudf.DataFrame()

df['key'] = ['a', 'b', 'c', 'd', 'e']
df['val'] = [7.6, 2.9, 7.1, 1.6, 2.2]

from blazingsql import BlazingContext
bc = BlazingContext(enable_progress_bar=True)

bc.create_table('game_1', df)

bc.sql('SELECT * FROM game_1 WHERE val > 4') # the query progress will be shown
```

| | Key | Value |
| - | -:| ---:|
| 0 | a | 7.6 |
| 1 | b | 7.1 |

Create and query a table from a AWS S3 bucket:

```python
from blazingsql import BlazingContext
bc = BlazingContext()

bc.s3('blazingsql-colab', bucket_name='blazingsql-colab')

bc.create_table('taxi', 's3://blazingsql-colab/yellow_taxi/taxi_data.parquet')

bc.sql('SELECT passenger_count, trip_distance FROM taxi LIMIT 2')
```

| | passenger_count | fare_amount |
| - | -:| ---:|
| 0 | 1.0 | 1.1 |
| 1 | 1.0 | 0.7 |

## Examples

| Notebook Title | Description | Try Now |
| -------------- | ----------- | ------- |
| Welcome Notebook | An introduction to BlazingSQL Notebooks and the GPU Data Science Ecosystem. | <a href='https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/welcome.ipynb'><img src="https://blazingsql.com/launch-notebooks.png" alt="Launch on BlazingSQL Notebooks" width="500"/></a> |
| The DataFrame | Learn how to use BlazingSQL and cuDF to create GPU DataFrames with SQL and Pandas-like APIs. | <a href='https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/intro_notebooks/the_dataframe.ipynb'><img src="https://blazingsql.com/launch-notebooks.png" alt="Launch on BlazingSQL Notebooks" width="500"/></a> |
| Data Visualization | Plug in your favorite Python visualization packages, or use GPU accelerated visualization tools to render millions of rows in a flash. | <a href='https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/intro_notebooks/data_visualization.ipynb'><img src="https://blazingsql.com/launch-notebooks.png" alt="Launch on BlazingSQL Notebooks" width="500"/></a> |
| Machine Learning | Learn about cuML, mirrored after the Scikit-Learn API, it offers GPU accelerated machine learning on GPU DataFrames. | <a href='https://app.blazingsql.com/jupyter/user-redirect/lab/workspaces/auto-b/tree/Welcome_to_BlazingSQL_Notebooks/intro_notebooks/machine_learning.ipynb'><img src="https://blazingsql.com/launch-notebooks.png" alt="Launch on BlazingSQL Notebooks" width="500"/></a> |

## Documentation
You can find our full documentation at [docs.blazingdb.com](https://docs.blazingdb.com/docs).

# Prerequisites
* [Anaconda or Miniconda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html) installed
* OS Support
  * Ubuntu 16.04/18.04 LTS
  * CentOS 7
* GPU Support
  * Pascal or Better
  * Compute Capability >= 6.0
* CUDA Support
  * 11.0
  * 11.2
* Python Support
  * 3.7
  * 3.8
# Install Using Conda
BlazingSQL can be installed with conda ([miniconda](https://conda.io/miniconda.html), or the full [Anaconda distribution](https://www.anaconda.com/download)) from the [blazingsql](https://anaconda.org/blazingsql/) channel:

## Stable Version
```bash
conda install -c blazingsql -c rapidsai -c nvidia -c conda-forge -c defaults blazingsql python=$PYTHON_VERSION cudatoolkit=$CUDA_VERSION
```
Where $CUDA_VERSION is 10.1, 10.2 or 11.0  and $PYTHON_VERSION is 3.7 or 3.8
*For example for CUDA 10.1 and Python 3.7:*
```bash
conda install -c blazingsql -c rapidsai -c nvidia -c conda-forge -c defaults blazingsql python=3.7 cudatoolkit=10.1
```

## Nightly Version
For nightly version cuda 11+ are only supported, see https://github.com/rapidsai/cudf#cudagpu-requirements
```bash
conda install -c blazingsql-nightly -c rapidsai-nightly -c nvidia -c conda-forge -c defaults blazingsql python=$PYTHON_VERSION  cudatoolkit=$CUDA_VERSION
```
Where $CUDA_VERSION is 11.0 or 11.2 and $PYTHON_VERSION is 3.7 or 3.8
*For example for CUDA 11.2 and Python 3.8:*
```bash
conda install -c blazingsql-nightly -c rapidsai-nightly -c nvidia -c conda-forge -c defaults blazingsql python=3.8  cudatoolkit=11.2
```

# Build/Install from Source (Conda Environment)
This is the recommended way of building all of the BlazingSQL components and dependencies from source. It ensures that all the dependencies are available to the build process.

## Stable Version

### Install build dependencies
```bash
conda create -n bsql python=$PYTHON_VERSION
conda activate bsql
./dependencies.sh rapids=0.19 cuda=$CUDA_VERSION
```
Where $CUDA_VERSION is is 10.1, 10.2 or 11.0 and $PYTHON_VERSION is 3.7 or 3.8
*For example for CUDA 10.1 and Python 3.7:*
```bash
conda create -n bsql python=3.7
conda activate bsql
./dependencies.sh rapids=0.19 cuda=10.1
```

### Build
The build process will checkout the BlazingSQL repository and will build and install into the conda environment.

```bash
cd $CONDA_PREFIX
git clone https://github.com/BlazingDB/blazingsql.git
cd blazingsql
git checkout main
export CUDACXX=/usr/local/cuda/bin/nvcc
./build.sh
```
NOTE: You can do `./build.sh -h` to see more build options.

$CONDA_PREFIX now has a folder for the blazingsql repository.

## Nightly Version

### Install build dependencies
For nightly version cuda 11+ are only supported, see https://github.com/rapidsai/cudf#cudagpu-requirements
```bash
conda create -n bsql python=$PYTHON_VERSION
conda activate bsql
./dependencies.sh rapids=21.06 cuda=$CUDA_VERSION nightly
```
Where $CUDA_VERSION is 11.0 or 11.2 and $PYTHON_VERSION is 3.7 or 3.8
*For example for CUDA 11.2 and Python 3.8:*
```bash
conda create -n bsql python=3.8
conda activate bsql
./dependencies.sh rapids=21.06 cuda=11.2 nightly
```

### Build
The build process will checkout the BlazingSQL repository and will build and install into the conda environment.

```bash
cd $CONDA_PREFIX
git clone https://github.com/BlazingDB/blazingsql.git
cd blazingsql
export CUDACXX=/usr/local/cuda/bin/nvcc
./build.sh
```
NOTE: You can do `./build.sh -h` to see more build options.

NOTE: You can perform static analysis with cppcheck with the command `cppcheck  --project=compile_commands.json` in any of the cpp project build directories.

$CONDA_PREFIX now has a folder for the blazingsql repository.

#### Storage plugins
To build without the storage plugins (AWS S3, Google Cloud Storage) use the next arguments:
```bash
# Disable all storage plugins
./build.sh disable-aws-s3 disable-google-gs

# Disable AWS S3 storage plugin
./build.sh disable-aws-s3

# Disable Google Cloud Storage plugin
./build.sh disable-google-gs
```
NOTE: By disabling the storage plugins you don't need to install previously AWS SDK C++ or Google Cloud Storage (neither any of its dependencies).

#### SQL providers
To build without the SQL providers (MySQL, PostgreSQL, SQLite, Snowflake) use the next arguments:
```bash
# Disable all SQL providers
./build.sh disable-mysql disable-sqlite disable-postgresql

# Disable MySQL provider
./build.sh disable-mysql

# Disable Snowflake provider
./build.sh disable-snowflake
...
```
NOTES:
- By disabling the storage plugins you don't need to install mysql-connector-cpp=8.0.23 libpq=13 sqlite=3 (neither any of its dependencies).
- Currenlty we support only MySQL and Snowflake but PostgreSQL and SQLite will be ready for the next version!

#### SnowFlake
SnowFlake works with [unixODBC](http://www.unixodbc.com/). There are two options to connect to snowflake which are the following:

##### Option 1
Execute the script `./snowflake_odbc_setup.sh` which will install and configure [snowflake driver for odbc 2.23.2](https://sfc-repo.snowflakecomputing.com/odbc/linux/2.23.2/index.html)

After the installation now you can start using snowflake without creating an odbc.ini file, an example is shown below.

```python
from blazingsql import BlazingContext
bc = BlazingContext()
bc.create_table('MY_TABLE', 'MY_SNOWFLAKE_TABLE',
        from_sql='snowflake',
        server='MY_SNOWFLAKE_SERVER.snowflakecomputing.com',
        database='MY_SNOWFLAKE_DATABASE_NAME',
        schema='PUBLIC',
        username='MY_USER',
        password='MY_PASSWORD',
        table_batch_size=3000)
df = bc.sql("SELECT * FROM MY_TABLE")
print(df)
```
##### Option 2
Download the [snowflake driver for odbc 2.23.2](https://sfc-repo.snowflakecomputing.com/odbc/linux/2.23.2/index.html) and unpack it in the /opt directory
Create the odbc.ini and odbcinst.ini files in the /etc directory in order to register a DSN (Datasource provider)

```ini
# /etc/odbcinst.ini
[SnowflakeDSIIDriver]
APILevel=1
ConnectFunctions=YYY
Description=Snowflake DSII
Driver=/path/to/snowflake_odbc/lib/libSnowflake.so
DriverODBCVer=03.52
SQLLevel=1

# /etc/odbc.ini
[MyDSN]
Description=SnowflakeDB
Driver=SnowflakeDSIIDriver
Locale=en-US
SERVER=myaccount.snowflakecomputing.com
PORT=443
SSL=on
ACCOUNT=myaccount
```
NOTE: Don't forget to update the odbc.ini and odbcinst.ini shown above with snowflake connection parameters

After this configuration you can now connect to snowflake. You can use this code snippet as an example
```python
from blazingsql import BlazingContext
bc = BlazingContext()
bc.create_table('MY_TABLE', 'MY_SNOWFLAKE_TABLE',
        from_sql='snowflake',
        dsn='MyDSN',
        database='MY_SNOWFLAKE_DATABASE_NAME',
        schema='PUBLIC',
        username='MY_USER',
        password='MY_PASSWORD',
        table_batch_size=3000)
df = bc.sql("SELECT * FROM MY_TABLE")
print(df)
```

# Documentation
User guides and public APIs documentation can be found at [here](https://docs.blazingdb.com/docs)

Our internal code architecture can be built using Spinx.
```bash
conda install -c conda-forge doxygen
cd $CONDA_PREFIX
cd blazingsql/docsrc
pip install -r requirements.txt
make doxygen
make html
```
The generated documentation can be viewed in a browser at `blazingsql/docsrc/build/html/index.html`


# Community
## Contributing
Have questions or feedback? Post a [new github issue](https://github.com/blazingdb/blazingsql/issues/new/choose).

Please see our [guide for contributing to BlazingSQL](CONTRIBUTING.md).

## Contact
Feel free to join our channel (#blazingsql) in the RAPIDS-GoAi Slack: [![join RAPIDS-GoAi workspace](https://badgen.net/badge/slack/RAPIDS-GoAi/purple?icon=slack)](https://join.slack.com/t/rapids-goai/shared_invite/enQtMjE0Njg5NDQ1MDQxLTJiN2FkNTFkYmQ2YjY1OGI4NTc5Y2NlODQ3ZDdiODEwYmRiNTFhMzNlNTU5ZWJhZjA3NTg4NDZkMThkNTkxMGQ).

You can also email us at [info@blazingsql.com](info@blazingsql.com) or find out more details on [BlazingSQL.com](https://blazingsql.com).

## License
[Apache License 2.0](LICENSE)

## RAPIDS AI - Open GPU Data Science

The RAPIDS suite of open source software libraries aim to enable execution of end-to-end data science and analytics pipelines entirely on GPUs. It relies on NVIDIA® CUDA® primitives for low-level compute optimization, but exposing that GPU parallelism and high-bandwidth memory speed through user-friendly Python interfaces.

## Apache Arrow on GPU

The GPU version of [Apache Arrow](https://arrow.apache.org/) is a common API that enables efficient interchange of tabular data between processes running on the GPU. End-to-end computation on the GPU avoids unnecessary copying and converting of data off the GPU, reducing compute time and cost for high-performance analytics common in artificial intelligence workloads. As the name implies, cuDF uses the Apache Arrow columnar data format on the GPU. Currently, a subset of the features in Apache Arrow are supported.


CHANGELOG.md
# BlazingSQL 21.06.00 (Date TBS)


## New Features
- #1471 Unbounded partitioned windows 
- #1445 Support for CURRENT_DATE, CURRENT_TIME and CURRENT_TIMESTAMP
- #1505 Support for right outer join
- #1523 Support for DURATION type
- #1552 Support for concurrency in E2E tests

## Improvements
- #1464 Better Support for unsigned types in C++ side
- #1511 Folder refactoring related to caches, kernels, execution_graph, BlazingTable
- #1522 Improve data loading when the algebra contains only BindableScan/Scan and Limit
- #1524 Enable support for spdlog 1.8.5
- #1547 Update RAPIDS version references
- #1539 Support ORDERing by null values
- #1551 Support for spdlog 1.8.5
- #1553 multiple columns inside COUNT() statement

## Bug Fixes
- #1455 Support for IS NOT FALSE condition
- #1502 Fix IS NOT DISTINCT FROM with joins
- #1475 Fix wrong results from timestampdiff/add
- #1528 Fixed build issues due to cudf aggregation API change 
- #1540 Comparing param set to true for e2e
- #1543 Enables provider unit_tests
- #1548 Fix orc statistic building
- #1550 Fix Decimal/Fixed Point issue
- #1519 Fix for max_bytes_chunk_read param to csv files
- #1559 Fix `ucx-py` versioning specs
- #1557 Reading chunks of max bytes for csv files


# BlazingSQL 0.19.0 (April 21, 2021)

## New Features
- #1367 OverlapAccumulator Kernel
- #1364 Implement the concurrent API (bc.sql with token, bc.status, bc.fetch)
- #1426 Window Functions without partitioning 
- #1349 Add e2e test for Hive Partitioned Data
- #1396 Create tables from other RDBMS
- #1427 Support for CONCAT alias operator
- #1424 Add get physical plan with explain
- #1472 Implement predicate pushdown for data providers

## Improvements
- #1325 Refactored CacheMachine.h and CacheMachine.cpp 
- #1322 Updated and enabled several E2E tests
- #1333 Fixing build due to cudf update
- #1344 Removed GPUCacheDataMetadata class
- #1376 Fixing build due to some strings refactor in cudf, undoing the replace workaround
- #1430 Updating GCP to >= version
- #1331 Added flag to enable null e2e testing
- #1418 Adding support for docker image
- #1434 Added documentation for C++ and Python in Sphinx
- #1419 Added concat cache machine timeout 
- #1444 Updating GCP to >= version
- #1349 Add e2e test for Hive Partitioned Data
- #1447 Improve getting estimated output num rows
- #1433 Adding ArrowCacheData, refactoring CacheData files
- #1473 Added Warning to Window Functions
- #1473 Added Warning to Window Functions
- #1482 Improve test script for blazingsql-testing-file
- #1480 Improve dependencies script
- #1433 Adding ArrowCacheData, refactoring CacheData files


## Bug Fixes
- #1335 Fixing uninitialized var in orc metadata and handling the parseMetadata exceptions properly
- #1339 Handling properly the nulls in case conditions with strings
- #1346 Delete allocated host chunks
- #1348 Capturing error messages due to exceptions properly
- #1350 Fixed bug where there are no projects in a bindable table scan
- #1359 Avoid cuda issues when free pinned memory
- #1365 Fixed build after sublibs changes on cudf
- #1369 Updated java path for powerpc build 
- #1371 Fixed e2e settings
- #1372 Recompute `columns_to_hash` in DistributeAggregationKernel
- #1375 Fix empty row_group_ids for parquet
- #1380 Fixed issue with int64 literal values 
- #1379 Remove ProjectRemoveRule
- #1389 Fix issue when CAST a literal
- #1387 Skip getting orc metadata for decimal type
- #1392 Fix substrings with nulls
- #1398 Fix performance regression
- #1401 Fix support for minus unary operation
- #1415 Fixed bug where num_batches was not getting set in BindableTableScan 
- #1413 Fix for null tests 13 and 23 of windowFunctionTest
- #1416 Fix full join when both tables contains nulls
- #1423 Fix temporary directory for hive partition test
- #1351 Fixed 'count distinct' related issues
- #1425 Fix for new joins API
- #1400 Fix for Column aliases when exists a Join op
- #1456 Raising exceptions on Python side for RAL
- #1466 SQL providers: update README.md
- #1470 Fix pre compiler flags for sql parsers
- #1504 Fixing some conflicts in Dockerfile

## Deprecated Features
- #1394 Disabled support for outer joins with inequalities 

# BlazingSQL 0.18.0 (February 24, 2021)

## New Features
- #1139 Adding centralized task executor for kernels
- #1200 Implement string REGEXP_REPLACE
- #1237 Added task memory management
- #1244 Added memory monitor ability to downgrade task data
- #1232 Update PartwiseJoin and JoinPartition kernel using the task executor internally
- #1238 Implements MergeStramKernel executor model
- #1259 Implements SortAndSamplernel executor model, also avoid setting up num of samples
- #1271 Added Hive utility for partitioned data
- #1289 Multiple concurrent query support 
- #1285 Infer PROTOCOL when Dask client is passed
- #1294 Add config options for logger
- #1301 Added usage of pinned buffers for communication and fixes various UCX related bugs
- #1298 Implement progress bar for run query (using tqdm)
- #1284 Initial support for Windows Function
- #1303 Add support for INITCAP
- #1313 getting and using ORC metadata
- #1347 Fixing issue when reading orc metadata from DATE dtype
- #1338 Window Function support for LEAD and LAG statements 
- #1362 give useful message when file extension is not recognized
- #1361 Supporting first_value and last_value for Window Function


## Improvements
- #1293 Added optional acknowledgments to message sending
- #1236 Moving code from header files to implementation files
- #1257 Expose the reset max memory usage C++ API to python
- #1256 Improve Logical project documentation
- #1262 Stop depending on gtest for runtime
- #1261 Improve storage plugin output messages
- #1153 Enable warnings and fixes
- #1267 Added retrys to comms, fixed deadlocks in executor and order by. Improved logging and error management. Caches have names. Improved Joins
- #1239 Reducing Memory pressure by moving shuffle data to cpu before transmission
- #1278 Fix race conditions with UCX

dependencies.sh
#!/bin/bash
#
# This script install BlazingSQL dependencies based on rapids version
#
set -e

NUMARGS=$#
ARGS=$*

VALIDARGS="rapids cuda -h nightly all e2e mysql postgres snowflake"

HELP="$0 [-h] rapids=[rapids version] cuda=[cuda version]
  all         - Will install e2e mysql postgres and snowflake dependencies
  e2e         - Install e2e basic dependencies for e2e tests
  -h          - Print help text
  mysql       - Install mysql dependencies for e2e tests
  nightly     - Install nightly dependencies
  postgres    - Install postgres dependencies for e2e tests
  snowflake   - Install snowflake dependencies for e2e tests
"

export GREEN='\033[0;32m'
export RED='\033[0;31m'
BOLDGREEN="\e[1;${GREEN}"
ITALICRED="\e[3;${RED}"
ENDCOLOR="\e[0m"

RAPIDS_VERSION="21.06"
UCX_PY_VERSION="0.20"
CUDA_VERSION="11.0"
CHANNEL=""

function hasArg {
    (( ${NUMARGS} != 0 )) && (echo " ${ARGS} " | grep -q " $1 ")
}

if hasArg -h; then
    echo "${HELP}"
    exit 0
fi

# Check for valid usage
if (( ${NUMARGS} != 0 )); then
    for a in ${ARGS}; do
      if [[ $a == "rapids="* ]]; then
          RAPIDS_VERSION=${a#"rapids="}
          if [ -z $RAPIDS_VERSION ] || [ $RAPIDS_VERSION == $a ] ; then
              echo "Invalid option: ${a}"
              exit 1
          fi
      fi
      if [[ $a == "cuda="* ]]; then
          CUDA_VERSION=${a#"cuda="}
          if [ -z $CUDA_VERSION ] || [ $CUDA_VERSION == $a ] ; then
              echo "Invalid option: ${a}"
              exit 1
          fi
      fi
      a=${a%=*}
      if ! (echo " ${VALIDARGS} " | grep -q " ${a} "); then
          echo "Invalid option: ${a}"
          exit 1
      fi
    done
else
    echo "USAGE: ${HELP}"
fi

if hasArg nightly; then
  CHANNEL="-nightly"
fi

if [ "$RAPIDS_VERSION" ] && [ "$CUDA_VERSION" ]; then
  echo -e "${GREEN}Installing build dependencies${ENDCOLOR}"
  conda install --yes -c conda-forge spdlog'>=1.7.0,<2.0.0a0' google-cloud-cpp=1.25 ninja mysql-connector-cpp=8.0.23 libpq=13 nlohmann_json=3.9.1 unixodbc=2.3.9
  # NOTE cython and cmake must be the same of cudf (for 0.11 and 0.12 cython is >=0.29,<0.30)
  conda install --yes -c conda-forge cmake>=3.18 gtest==1.10.0=h0efe328_4 gmock cppzmq cython=0.29 openjdk=8.0 maven jpype1 netifaces pyhive pytest tqdm ipywidgets boost-cpp=1.76.0

  echo -e "${GREEN}Install RAPIDS dependencies${ENDCOLOR}"
  conda install --yes -c rapidsai$CHANNEL -c nvidia -c conda-forge -c defaults dask-cuda=$RAPIDS_VERSION dask-cudf=$RAPIDS_VERSION cudf=$RAPIDS_VERSION ucx-py=$UCX_PY_VERSION ucx-proc=*=gpu cudatoolkit=$CUDA_VERSION
fi

if hasArg all || hasArg e2e; then
  echo -e "${GREEN}Install E2E test dependencies${ENDCOLOR}"
  conda install --yes -c conda-forge openjdk=8.0 maven pyspark=3.0.0 pytest
  pip install pydrill openpyxl pymysql gitpython pynvml gspread oauth2client docker 'sql-metadata==1.12.0' pyyaml yamale
fi

if hasArg all || hasArg mysql; then
  echo -e "${GREEN}Install mysql test dependencies${ENDCOLOR}"
  conda install --yes -c conda-forge mysql-connector-python
fi

if hasArg all || hasArg postgres; then
  echo -e "${GREEN}Install postgres test dependencies${ENDCOLOR}"
  conda install --yes -c conda-forge psycopg2
fi

if hasArg all || hasArg snowflake; then
  echo -e "${GREEN}Install snowflake test dependencies${ENDCOLOR}"
  pip install snowflake-connector-python==2.4.2
fi
if [ $? -eq 0 ]; then
  echo -e "${GREEN}Installation complete${ENDCOLOR}"
else
  echo -e "${RED}Installation failed${ENDCOLOR}"
fi
